### PR TITLE
SONARIAC-542 Adapt ExecForm implementation

### PR DESCRIPTION
**Pull request by rudy-regazzoni-sonarsource**

### CODE REVIEW

**Comment by rudy-regazzoni-sonarsource** [comment]
Baby step to move Shell/Exec/Heredoc form to Argument.
Focus on ExecForm, had to introduce a method `tokens()` in Argument that may not be needed anymore once the move is complete.
Started the deprecation of  `literals()`, which will be removed completely at the end of the move.

**Review by nils-werner-sonarsource — CHANGES_REQUESTED** [review]
You can remove some technical dept, by removing ExecFromLiteral. Reduce the complexity of the grammar I commented on.

Maybe the ExecForm and ExecFormUtils can be a bit more generalized. So that we can use a utils class in all checks which accept a command structure. So there is no difference in handling exec and shell. However, this would already be another baby step.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormLiteralImpl.java [hunk 1]
```diff
@@ -22,13 +22,13 @@
 import java.util.List;
 import org.sonar.iac.common.api.tree.Tree;
 import org.sonar.iac.docker.tree.api.ExecFormLiteral;
-import org.sonar.iac.docker.tree.api.SyntaxToken;
+import org.sonar.iac.docker.tree.api.ExpandableStringLiteral;
 
 public class ExecFormLiteralImpl extends AbstractDockerTreeImpl implements ExecFormLiteral {
```
[the remark this answers is not in the document]
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormLiteralImpl.java:27]: Damn that's a very good point.
I removed this useless layer, this simplify things a bit.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java [hunk 2]
```diff
@@ -260,22 +258,19 @@ public HereDocument hereDocument(SyntaxToken content) {
...
-        for (Tuple<SyntaxToken, SyntaxToken> comaAndLiteral : comaAndLiterals) {
-          separators.add(comaAndLiteral.first());
-          elements.add(new ExecFormLiteralImpl(comaAndLiteral.second()));
+    SeparatedList<ExpandableStringLiteral> separatedList = new SeparatedListImpl<>(elements, separators);
+    if (firstExpression.isPresent()) {
+      elements.add(firstExpression.get());
+      if (otherExpressions.isPresent()) {
+        for (Tuple<SyntaxToken, ExpandableStringLiteral> expressionWithComma : otherExpressions.get()) {
+          separators.add(expressionWithComma.first());
+          elements.add(expressionWithComma.second());
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java:273]: Much better!

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/ExecFormTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.sonar.iac.docker.tree.api.Argument;
+import org.sonar.iac.docker.tree.api.DockerTree;
+import org.sonar.iac.docker.tree.api.ExpandableStringCharacters;
+import org.sonar.iac.docker.tree.api.ExpandableStringLiteral;
+import org.sonar.iac.docker.tree.api.Expression;
+import org.sonar.iac.docker.tree.api.Literal;
+
+public class ExecFormTestUtils {
+
+  public static String toString(Argument arg) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/test/java/org/sonar/iac/docker/tree/impl/ExecFormTestUtils.java:32]: 1. Please do not use the method name `toString` it is reserved for other purposes
2. You can simply use ArgumentUtils instead of duplicating its logic.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormImpl.java [hunk 4]
```diff
@@ -58,11 +59,21 @@ public SyntaxToken leftBracket() {
...
-    for (ExecFormLiteral element : literalsWithSeparators.elements()) {
-      result.add(element.value());
+    throw new UnsupportedOperationException("Method to be removed from LiteralList interface once SONARIAC-541 and SONARIAC-572 are done.");
+  }
+
+  @Override
+  public List<Argument> arguments() {
+    List<Argument> result = new ArrayList<>();
+    for (ExpandableStringLiteral element : expressionsWithSeparators.elements()) {
+      result.add(new ArgumentImpl(List.of(element)));
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormImpl.java:76]: The tree class should not take care of this. Maybe we can let the grammar provide the arguments in the separate list directly? WDYT? (However, we can do it in another PR)

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM! Still a bit odd due to all the deprecation and replacements. We should get rid of the LiteralList as well (Next step) This can by simplified to List<Argument>.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormImpl.java (hunk 4 above)
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/tree/impl/ExecFormImpl.java]: I agree on this, I put this in place in my last commit.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java [hunk 5]
```diff
@@ -262,22 +260,23 @@ public HereDocument hereDocument(SyntaxToken content) {
     return new HereDocumentImpl(content);
   }
 
-  public ExecForm execForm(SyntaxToken leftBracket,
-    Optional<Tuple<SyntaxToken, Optional<List<Tuple<SyntaxToken, SyntaxToken>>>>> literals,
+  public Argument expressionAsArgument(Expression expression) {
```
Comment (nils-werner-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/parser/TreeFactory.java:263]:
```suggestion
  public Argument singleExpressionArguement(Expression expression) {
```

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM! Can you explain, why we loose so many issues on ITs?

**Comment by rudy-regazzoni-sonarsource** [comment]
> LGTM! Can you explain, why we loose so many issues on ITs?

This is because we already have a rule using Exec/Shell form, working with the contract that the method `literals()` will provide expected elements (it use it directly and indirectly with `srcs()`/`dest()`).
This is still working for Shell/Heredoc form, but now `ExecForm` throw the `UnsupportedOperationException`, which throw error during rule execution and prevent further scanning / issue detection.

This is why at first I tried to keep the previous `literals()` method of `ExecForm` working temporarily, but at some point it was bringing too much hassle.

So unfortunately we can only expect to detect those issues back once we are done with the transition on ExecForm/ShellForm/HeredocForm + removed old method `literals()` and adapted the rule to `arguments()`.

**Comment by nils-werner-sonarsource** [comment]
> So unfortunately we can only expect to detect those issues back once we are done with the transition on ExecForm/ShellForm/HeredocForm + removed old method `literals()` and adapted the rule to `arguments()`.

Okay. Another approach would be to use `arguments()` inside the `literals()` declaration. So it would prove that this rule is working with the new data from `arguments()`. Instead of throwing an exception, you could map the arguments to the expected output of ` `literals()`

**Review by rudy-regazzoni-sonarsource — COMMENTED** [review]
LGTM !
